# A lady wishes to color her fingernails on one hand using at most two of the colors [red, yellow, blue]

# How many ways can she do this?

___
### If we choose two colors, they can be applied to her fingers $2^{5}$ different ways

### But there are $\binom{3}{2}$ possible ways to choose the two colors

## So, the total number of ways she can do this is $\binom{3}{2}2^{5} = 3\cdot32 = 96$

# NOT DONE YET: we've double counted for some of the possibilities

### Let's say we choose red and yellow

### Then, there are two possible applications of the $2^{5}$ possibilities that are a single color - all red, and all yellow

### But if we choose red and blue, then there are also two single color possibilities - all red and all green

### Similarly, if we choose yellow and blue, then there's all yellow and all blue

## So when we count the 96 possibilities, we're counting the pure color possibilities twice each, and therefore we need to remove 3

# So, the total number of possibilities is equal to 96-3 = 96

___
We can prove this by listing out all possible combinations

In [1]:
import pandas as pd

In [7]:
df = pd.DataFrame(columns = range(1,6))

colors = ['R','Y','B']

i = 0
for finger_1 in colors:
    for finger_2 in colors:
        for finger_3 in colors:
            for finger_4 in colors:
                for finger_5 in colors:
                    df.loc[i] = [finger_1, finger_2, finger_3, finger_4, finger_5]
                    i+=1

In [8]:
df

1  2  3  4  5
0    R  R  R  R  R
1    R  R  R  R  Y
2    R  R  R  R  B
3    R  R  R  Y  R
4    R  R  R  Y  Y
5    R  R  R  Y  B
6    R  R  R  B  R
7    R  R  R  B  Y
8    R  R  R  B  B
9    R  R  Y  R  R
10   R  R  Y  R  Y
11   R  R  Y  R  B
12   R  R  Y  Y  R
13   R  R  Y  Y  Y
14   R  R  Y  Y  B
15   R  R  Y  B  R
16   R  R  Y  B  Y
17   R  R  Y  B  B
18   R  R  B  R  R
19   R  R  B  R  Y
20   R  R  B  R  B
21   R  R  B  Y  R
22   R  R  B  Y  Y
23   R  R  B  Y  B
24   R  R  B  B  R
25   R  R  B  B  Y
26   R  R  B  B  B
27   R  Y  R  R  R
28   R  Y  R  R  Y
29   R  Y  R  R  B
..  .. .. .. .. ..
213  B  Y  B  B  R
214  B  Y  B  B  Y
215  B  Y  B  B  B
216  B  B  R  R  R
217  B  B  R  R  Y
218  B  B  R  R  B
219  B  B  R  Y  R
220  B  B  R  Y  Y
221  B  B  R  Y  B
222  B  B  R  B  R
223  B  B  R  B  Y
224  B  B  R  B  B
225  B  B  Y  R  R
226  B  B  Y  R  Y
227  B  B  Y  R  B
228  B  B  Y  Y  R
229  B  B  Y  Y  Y
230  B  B  Y  Y  B
231  B  B  Y  B  R
232  B  B  Y  B  Y
233  B  B  Y  B  B
234  B  B  B  R  R
235  B  B  B  R  Y
236  B  B  B  R  B
237  B  B  B  Y  R
238  B  B  B  Y  Y
239  B  B  B  Y  B
240  B  B  B  B  R
241  B  B  B  B  Y
242  B  B  B  B  B

[243 rows x 5 columns]

In [9]:
len(df)

243

## Our loop generated 243 rows, which is equal to $3^{5}$, so that's the correct number of combinations

### Now, we count only the rows that have 2 colors or fewer

In [18]:
df2 = df[df.apply(lambda x: len(x.drop_duplicates()), axis = 1)<=2]
df2

1  2  3  4  5
0    R  R  R  R  R
1    R  R  R  R  Y
2    R  R  R  R  B
3    R  R  R  Y  R
4    R  R  R  Y  Y
6    R  R  R  B  R
8    R  R  R  B  B
9    R  R  Y  R  R
10   R  R  Y  R  Y
12   R  R  Y  Y  R
13   R  R  Y  Y  Y
18   R  R  B  R  R
20   R  R  B  R  B
24   R  R  B  B  R
26   R  R  B  B  B
27   R  Y  R  R  R
28   R  Y  R  R  Y
30   R  Y  R  Y  R
31   R  Y  R  Y  Y
36   R  Y  Y  R  R
37   R  Y  Y  R  Y
39   R  Y  Y  Y  R
40   R  Y  Y  Y  Y
54   R  B  R  R  R
56   R  B  R  R  B
60   R  B  R  B  R
62   R  B  R  B  B
72   R  B  B  R  R
74   R  B  B  R  B
78   R  B  B  B  R
..  .. .. .. .. ..
164  B  R  R  R  B
168  B  R  R  B  R
170  B  R  R  B  B
180  B  R  B  R  R
182  B  R  B  R  B
186  B  R  B  B  R
188  B  R  B  B  B
202  B  Y  Y  Y  Y
203  B  Y  Y  Y  B
205  B  Y  Y  B  Y
206  B  Y  Y  B  B
211  B  Y  B  Y  Y
212  B  Y  B  Y  B
214  B  Y  B  B  Y
215  B  Y  B  B  B
216  B  B  R  R  R
218  B  B  R  R  B
222  B  B  R  B  R
224  B  B  R  B  B
229  B  B  Y  Y  Y
230  B  B  Y  Y  B
232  B  B  Y  B  Y
233  B  B  Y  B  B
234  B  B  B  R  R
236  B  B  B  R  B
238  B  B  B  Y  Y
239  B  B  B  Y  B
240  B  B  B  B  R
241  B  B  B  B  Y
242  B  B  B  B  B

[93 rows x 5 columns]

In [19]:
len(df2)

93

# 93! Not 96